### PR TITLE
Linkname removal (step 1)

### PR DESCRIPTION
**Pull request by vmg**
## Description

First step towards fixing https://github.com/vitessio/vitess/issues/16015.

In this PR I've cleaned up to `Atof` routines, the internal `strhash` and the memory size calculator. The only remaining `linkname` usage in the `hack` package is to disable serialization randomness in ProtoBuf.

## Related Issue(s)

- https://github.com/vitessio/vitess/issues/16015

## Checklist

-   [ ] "Backport to:" labels have been added if this change should be back-ported to release branches
-   [ ] If this change is to be back-ported to previous releases, a justification is included in the PR description
-   [ ] Tests were added or are not required
-   [ ] Did the new or modified tests pass consistently locally and on CI?
-   [ ] Documentation was added or is not required

### CODE REVIEW

**Review by dbussink — APPROVED** [review]
A bunch of copy paste from Go, but that seems like the right answer here!

File: go/hack/msize.go [hunk 1]
```diff
@@ -0,0 +1,76 @@
...
+THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+(INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+*/
+
+// Copied from the Go runtime, msize_noallocheaders.go
+
+package hack
+
+const (
```
Comment (frouioui) [inline go/hack/msize.go:35]: Does that mean that every time the go runtime changes for this file (and also the package `strconv`) we will need to re-copy the file?

Comment (vmg) [reply go/hack/msize.go:35]: For `msize` yes, that's the price of keeping our memory usage statistics accurate. FWIW, this file hasn't been changed in more than 5 years: the sizes of the GC allocations are pretty much settled at this point.

For `strconv`, no. We have adapted the function to parse exactly the syntax that MySQL expects, so we do not need to ever upgrade it in the future.

Comment (frouioui) [reply go/hack/msize.go:35]: Okay got it. I was worried these were changing often every x release, but if it has not for over 5 years that's nice

**Review by frouioui — APPROVED** [review]

**Pull request merged**